base/parse-type.jl

```diff
@@ -0,0 +1,164 @@
+# Parses a string into a Julia type object, e.g. `Int`, `Array{Int, 2}`, etc.
+function Base.parse(::Type{T}, str::AbstractString) where T<:Type
+    # tryparse(Type) instead of (T), so we only have to do the type check once.
+    v = Base.tryparse(Type, str)
+    if v === nothing
+        throw(ArgumentError("invalid type expression: $str"))
+    end
+    # Don't pay for the assertion if not needed (~2 μs)
+    T === Type && return v
+    return v::T
+end
+function Base.tryparse(::Type{T}, str::AbstractString) where {T<:Type}
+    ast = Meta.parse(str, raise=false)
+    # (This only needs to be checked once, so it's checked here, rather than in
+    # _try_parse_type, since _try_parse_type is self-recursive.)
+    if ast isa Expr && ast.head === :incomplete
+        return nothing
+    end
+    v = _try_parse_type(ast, false)
+    v === nothing && return nothing
+    # Don't pay for the type check if not needed (~2 μs)
+    T === Type && return v
+    v isa T || return nothing
+    return v
+end
+
+# Simplify the commonly repeated pattern checking return values.
+macro _or_nothing(expr)
+    Base.remove_linenums!(quote
+        v = $(esc(expr))
+        v === nothing && return nothing
+        v
+    end)
+end
```

Comment:
What does this accomplish compared to just `return v`?

Reply:
It's because I don't want to _return_ if `v` is not nothing; i want to continue. I should have called this `@_bail_if_nothing`, since if the value is nothing, it will _return early_, rather than continuing the function.

I'll rename it, and then can you let me know if this seems acceptable?